src/librbd/cache/ReplicatedWriteLog.cc

```diff
@@ -0,0 +1,5130 @@
+    }
+    return os;
+  };
+
+  C_DiscardRequest(T &rwl, const utime_t arrived, Extents &&image_extents,
+		   const int skip_partial_discard, Context *user_req)
+    : C_BlockIORequest<T>(rwl, arrived, std::move(image_extents), bufferlist(), 0, user_req),
+    m_skip_partial_discard(skip_partial_discard) {
+    if (RWL_VERBOSE_LOGGING) {
+      ldout(rwl.m_image_ctx.cct, 99) << this << dendl;
+    }
+  }
+
+  ~C_DiscardRequest() {
+    if (RWL_VERBOSE_LOGGING) {
+      ldout(rwl.m_image_ctx.cct, 99) << this << dendl;
+    }
+  }
+
+  template <typename... U>
+  static inline std::shared_ptr<C_DiscardRequest<T>> create(U&&... arg)
+  {
+    return SharedPtrContext::create<C_DiscardRequest<T>>(std::forward<U>(arg)...);
+  }
+
+  auto shared_from_this() {
+    return shared_from(this);
+  }
+
+  void finish_req(int r) {}
+
+  bool alloc_resources() override;
+
+  void deferred_handler() override { }
+
+  void dispatch() override;
+
+  const char *get_name() const override {
+    return "C_DiscardRequest";
+  }
+};
+
+/**
+ * This is the custodian of the BlockGuard cell for this compare and write. The
+ * block guard is acquired before the read begins to guarantee atomicity of this
+ * operation.  If this results in a write, the block guard will be released
+ * when the write completes to all replicas.
+ */
+template <typename T>
+struct C_CompAndWriteRequest : public C_WriteRequest<T> {
+  using C_BlockIORequest<T>::rwl;
+  bool m_compare_succeeded = false;
+  uint64_t *m_mismatch_offset;
+  bufferlist m_cmp_bl;
+  bufferlist m_read_bl;
+  friend std::ostream &operator<<(std::ostream &os,
+				  const C_CompAndWriteRequest<T> &req) {
+    os << (C_WriteRequest<T>&)req
+       << "m_cmp_bl=" << req.m_cmp_bl << ", "
+       << "m_read_bl=" << req.m_read_bl << ", "
+       << "m_compare_succeeded=" << req.m_compare_succeeded << ", "
+       << "m_mismatch_offset=" << req.m_mismatch_offset;
+    return os;
+  };
+
+  C_CompAndWriteRequest(T &rwl, const utime_t arrived, Extents &&image_extents,
+			bufferlist&& cmp_bl, bufferlist&& bl, uint64_t *mismatch_offset,
+			int fadvise_flags, Context *user_req)
+    : C_WriteRequest<T>(rwl, arrived, std::move(image_extents), std::move(bl), fadvise_flags, user_req),
+    m_mismatch_offset(mismatch_offset), m_cmp_bl(std::move(cmp_bl)) {
+    if (RWL_VERBOSE_LOGGING) {
+      ldout(rwl.m_image_ctx.cct, 99) << this << dendl;
+    }
+  }
+
+  ~C_CompAndWriteRequest() {
+    if (RWL_VERBOSE_LOGGING) {
+      ldout(rwl.m_image_ctx.cct, 99) << this << dendl;
+    }
+  }
+
+  template <typename... U>
+  static inline std::shared_ptr<C_CompAndWriteRequest<T>> create(U&&... arg)
+  {
+    return SharedPtrContext::create<C_CompAndWriteRequest<T>>(std::forward<U>(arg)...);
+  }
+
+  auto shared_from_this() {
+    return shared_from(this);
+  }
+
+  void finish_req(int r) override {
+    if (m_compare_succeeded) {
+      C_WriteRequest<T>::finish_req(r);
+    } else {
+      utime_t now = ceph_clock_now();
+
```

Comment:
I don't think we need to check whether m_perfcounter is null here. During init, it is initialized and isn't checked in other places.